src/UKHO.ADDS.EFS.Builder.S100/Pipelines/Assemble/DownloadFilesNode.cs

```diff
@@ -58,36 +58,105 @@ private static IEnumerable<BatchDetails> SelectLatestBatchesByProductEditionAndU
+
+            foreach (var directoryPath in fileDirectoryPaths)
+            {
+                CreateDirectoryIfNotExists(directoryPath, createdDirectories);
+            }
 
-                    await using var outputFileStream = new FileStream(downloadPath, FileMode.Create, FileAccess.ReadWrite);
+            // Now download all files (all directories are guaranteed to exist)
+            foreach (var item in allFilesToProcess)
+            {
+                var directoryPath = GetDirectoryPathForFile(workSpaceRootPath, item.FileName, workSpaceSpoolDataSetFilesPath, workSpaceSpoolSupportFilesPath);
+                var downloadPath = Path.Combine(directoryPath, item.FileName);
 
-                    var streamResult = await _fileShareReadOnlyClient.DownloadFileAsync(batch.BatchId, fileName, outputFileStream, correlationId, FileSizeInBytes);
+                await using var outputFileStream = new FileStream(downloadPath, FileMode.Create, FileAccess.ReadWrite);
 
-                    if (streamResult.IsFailure(out var error, out var value))
-                    {
-                        LogFssDownloadFailed(batch, fileName, error, correlationId);
+                var streamResult = await _fileShareReadOnlyClient.DownloadFileAsync(
+                    item.Batch.BatchId, item.FileName, outputFileStream, correlationId, FileSizeInBytes);
 
-                        return NodeResultStatus.Failed;
-                    }
+                if (streamResult.IsFailure(out var error, out var value))
+                {
+                    LogFssDownloadFailed(item.Batch, item.FileName, error, correlationId);
+                    return NodeResultStatus.Failed;
                 }
             }
 
             return NodeResultStatus.Succeeded;
         }
 
+        private static void CreateDirectoryIfNotExists(string downloadPath, HashSet<string> createdDirectories = null)
+        {
+            // Early return if we've already created this directory
+            if (createdDirectories?.Contains(downloadPath) == true)
+            {
+                return;
+            }
+
+            // Create directory if it doesn't exist
+            if (!Directory.Exists(downloadPath))
+            {
+                Directory.CreateDirectory(downloadPath);
+            }
+
+            // Add to our tracking set if we're using one
+            createdDirectories?.Add(downloadPath);
+        }
+
+        private string GetDirectoryPathForFile(string workSpaceRootPath, string fileName,string workSpaceSpoolDataSetFilesPath, string workSpaceSpoolSupportFilesPath)
+        {
+            var extension = Path.GetExtension(fileName);
+
+            // Handle numeric extensions (.000 to .999)
+            if (extension is { Length: 4 } && int.TryParse(extension[1..], out var extNum) && extNum is >= 0 and <= 999)
+            {
+                if (fileName.Length >= 7)
+                {
+                    var folderName = fileName.Substring(3, 4);
```

Comment:
Add commented to clarify what part of the filename is being extracted and why also we can use named constants for more clarity and can be used across class.